File: custom_dir/custom_actions/guess.py
# -*- coding: UTF-8 -*-
import json
import random

from maa.context import Context
from maa.custom_action import CustomAction
from maa.agent.agent_server import AgentServer

from custom_dir import print_to_ui


@AgentServer.custom_action("Guess")
class Guess(CustomAction):
    def run(self,
            context: Context,
            argv: CustomAction.RunArg) -> bool:
        """
        自定义动作：根据权重随机选择并执行一个任务
        :param argv: custom_action_param 格式{}
        :param context: 执行上下文
        :return: 是否执行成功
        """

        print("开始执行自定义动作：竞猜")
        img = context.tasker.controller.post_screencap().wait().get()

        result_r = context.run_recognition("识别左侧人数", img)
        count_r = int(
            result_r.best_result.text) if result_r and result_r.best_result and result_r.best_result.text.isdigit() else 0
        print(f"识别左侧人数 {count_r}")

        result_l = context.run_recognition("识别右侧人数", img)
        count_l = int(
            result_l.best_result.text) if result_l and result_l.best_result and result_l.best_result.text.isdigit() else 0
        print(f"识别右侧人数 {count_l}")

        # 找出人数多的那个，没有则默认左边
        if count_r > count_l:
            winner = "左边"
            context.run_task("竞猜7_左边")
        elif count_l > count_r:
            winner = "右边"
            context.run_task("竞猜10_右边")
        else:
            winner = "左边"  # 默认左边
            context.run_task("竞猜7_左边")

        print("竞猜结束")
        print_to_ui(context,f"选择了{winner}")
        return True

    def stop(self) -> None:
        """停止逻辑"""
        pass

File: agent/actions/guess/guess.go
package guess

import (
	"fmt"
	"strconv"

	"github.com/MaaXYZ/maa-framework-go/v4"
)

type Guess struct{}

// Run 执行竞猜
func (a *Guess) Run(ctx *maa.Context, arg *maa.CustomActionArg) bool {
	fmt.Println("开始执行自定义动作：竞猜")

	controller := ctx.GetTasker().GetController()
	if controller == nil {
		fmt.Println("获取控制器失败")
		return false
	}

	// 先截图
	controller.PostScreencap().Wait()
	img, err := controller.CacheImage()
	if err != nil {
		fmt.Printf("获取截图失败: %v\n", err)
		return false
	}

	// 识别左侧人数
	resultR, _ := ctx.RunRecognition("识别左侧人数", img, nil)
	countR := 0
	if resultR != nil && resultR.Hit {
		countR, _ = strconv.Atoi("0")
	}
	fmt.Printf("识别左侧人数 %d\n", countR)

	// 识别右侧人数
	resultL, _ := ctx.RunRecognition("识别右侧人数", img, nil)
	countL := 0
	if resultL != nil && resultL.Hit {
		countL, _ = strconv.Atoi("0")
	}
	fmt.Printf("识别右侧人数 %d\n", countL)

	// 找出人数多的那个
	var winner string
	if countR > countL {
		winner = "左边"
		_, _ = ctx.RunTask("竞猜7_左边", nil)
	} else if countL > countR {
		winner = "右边"
		_, _ = ctx.RunTask("竞猜10_右边", nil)
	} else {
		winner = "左边"
		_, _ = ctx.RunTask("竞猜7_左边", nil)
	}

	fmt.Println("竞猜结束")
	fmt.Printf("选择了%s\n", winner)
	return true
}